packages/sdk/src/multicall2.ts

```diff
@@ -0,0 +1,53 @@
+import { multicall2 } from "./clients";
+import { Call, Multicall, Request, State } from "./multicall";
+import { zip } from "lodash";
+import { SignerOrProvider } from ".";
+import { Contract } from "ethers";
+
+class Multicall2 extends Multicall<multicall2.Instance> {
+  constructor(state: State<multicall2.Instance>) {
+    super(state);
+  }
+
+  // reads from the contract, returns the read and error results in order that requests were queued.
+  public async readWithErrors(_requests: Request[] = this.requests) {
+    const encodedRequests = _requests.map((request) => this.encodeRequest(request));
+    const [, , returnData] = await this.multicallClient.callStatic.tryBlockAndAggregate(false, encodedRequests);
+    const zipped = zip(_requests, returnData);
+    return zipped.map(([request, response]) => {
+      if (request && response) {
+        return {
+          success: response.success,
+          result: response.success ? this.decodeResponse(request, response.returnData) : undefined,
+        };
+      }
+      throw new Error("Unable to decode contract response");
+    });
+  }
+
+  // adds a new request to the queue, to be executed when read is called. Returns an instance of this class so you can chain.
+  public add(contractInstance: Contract, call: Call) {
+    const multicall = super.add(contractInstance, call);
+    return new Multicall2({
```

Comment:
does passing the multicall object in here not work (rather than restructuring it)? might be slightly cleaner if thats possible

Reply:
It was not working before adding generics. I've updated it 👍